File: Data Structures & Algorithms/surrounded-regions/submission-5.java
class Solution {
    public void solve(char[][] board) {
        if (board == null || board.length == 0) {
            return;
        }

        boolean[][] visited = new boolean[board.length][board[0].length];
        List<List<int[]>> regions = new ArrayList<>();

        for (int row = 0; row < board.length; row++) {
            for (int col = 0; col < board[0].length; col++) {
                if (!visited[row][col] && board[row][col] == 'O') {
                    List<int[]> region = bfs(
                        new int[] {row, col},
                        board,
                        visited
                    );

                    regions.add(region);
                }
            }
        }

        for (List<int[]> region : regions) {
            boolean touchesEdge = false;

            for (int[] cell : region) {
                int row = cell[0];
                int col = cell[1];

                if (isEdge(row, col, board)) {
                    touchesEdge = true;
                    break;
                }
            }

            if (!touchesEdge) {
                for (int[] cell : region) {
                    board[cell[0]][cell[1]] = 'X';
                }
            }
        }
    }

    public List<int[]> bfs(
        int[] start,
        char[][] board,
        boolean[][] visited
    ) {
        Deque<int[]> queue = new ArrayDeque<>();
        List<int[]> region = new ArrayList<>();

        queue.offer(start);
        visited[start[0]][start[1]] = true;

        int[][] directions = {
            {1, 0},
            {-1, 0},
            {0, 1},
            {0, -1}
        };

        int rows = board.length;
        int cols = board[0].length;

        while (!queue.isEmpty()) {
            int[] cell = queue.poll();
            int row = cell[0];
            int col = cell[1];

            region.add(new int[] {row, col});

            for (int[] direction : directions) {
                int newRow = row + direction[0];
                int newCol = col + direction[1];

                if (
                    newRow >= 0 &&
                    newRow < rows &&
                    newCol >= 0 &&
                    newCol < cols &&
                    !visited[newRow][newCol] &&
                    board[newRow][newCol] == 'O'
                ) {
                    visited[newRow][newCol] = true;
                    queue.offer(new int[] {newRow, newCol});
                }
            }
        }

        return region;
    }

    public boolean isEdge(int row, int col, char[][] board) {
        return (
            row == 0 ||
            row == board.length - 1 ||
            col == 0 ||
            col == board[0].length - 1
        );
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-5.py
from collections import deque
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        def bfs(start, visited, grid):
            queue = deque([start])
            visited.add(start)
            ROWS = len(grid)
            COLS = len(grid[0])
            DIRECTIONS = [(1,0), (-1,0), (0,1), (0,-1)]

            while queue:
                row, col = queue.popleft()

                for dr, dc in DIRECTIONS:
                    new_row, new_col = row + dr, col+dc 
                    if (0 <= new_row < ROWS and 0 <= new_col < COLS and (new_row, new_col) not in visited and grid[new_row][new_col] == "1"):
                        visited.add((new_row,new_col))
                        queue.append((new_row,new_col))
            
        
        visited = set()
        total = 0
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == "1"  and (row, col) not in visited:
                    bfs((row,col), visited, grid)
                    total += 1
        return total
        